// motion-v/dist/es/index.mjs
import { addScaleCorrector } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/projection/styles/scale-correction.mjs";
import { default as default2 } from "./components/LayoutGroup.vue.mjs";
import { useLayoutGroup } from "./components/use-layout-group.mjs";
import { injectLayoutGroup, injectMotion, provideLayoutGroup, provideMotion } from "./components/context.mjs";
import { useDragControls } from "./features/gestures/drag/use-drag-controls.mjs";
import { domMax } from "./features/dom-max.mjs";
import { domAnimation } from "./features/dom-animation.mjs";
import { default as default3 } from "./components/RowValue.vue.mjs";
import { default as default4 } from "./components/animate-presence/AnimatePresence.vue.mjs";
import { default as default5 } from "./components/motion-config/MotionConfig.vue.mjs";
import { MotionValue, motionValue, motionValue as motionValue2 } from "./external/.pnpm/motion-dom@12.5.0/external/motion-dom/dist/es/value/index.mjs";
import { cancelFrame, frame, frameData } from "./external/.pnpm/motion-dom@12.5.0/external/motion-dom/dist/es/frameloop/frame.mjs";
import { hover } from "./external/.pnpm/motion-dom@12.5.0/external/motion-dom/dist/es/gestures/hover.mjs";
import { isDragActive } from "./external/.pnpm/motion-dom@12.5.0/external/motion-dom/dist/es/gestures/drag/state/is-active.mjs";
import { press } from "./external/.pnpm/motion-dom@12.5.0/external/motion-dom/dist/es/gestures/press/index.mjs";
import { time } from "./external/.pnpm/motion-dom@12.5.0/external/motion-dom/dist/es/frameloop/sync-time.mjs";
import { invariant } from "./external/.pnpm/motion-utils@12.5.0/external/motion-utils/dist/es/errors.mjs";
import { noop } from "./external/.pnpm/motion-utils@12.5.0/external/motion-utils/dist/es/noop.mjs";
import { progress } from "./external/.pnpm/motion-utils@12.5.0/external/motion-utils/dist/es/progress.mjs";
import { animate, createScopedAnimate } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/animation/animate/index.mjs";
import { animateMini } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/animation/animators/waapi/animate-style.mjs";
import { scroll } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/render/dom/scroll/index.mjs";
import { scrollInfo } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/render/dom/scroll/track.mjs";
import { inView } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/render/dom/viewport/index.mjs";
import { anticipate } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/anticipate.mjs";
import { backIn, backInOut, backOut } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/back.mjs";
import { circIn, circInOut, circOut } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/circ.mjs";
import { cubicBezier } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/cubic-bezier.mjs";
import { easeIn, easeInOut, easeOut } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/ease.mjs";
import { mirrorEasing } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/modifiers/mirror.mjs";
import { reverseEasing } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/modifiers/reverse.mjs";
import { steps } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/easing/steps.mjs";
import { inertia } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/animation/generators/inertia.mjs";
import { keyframes } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/animation/generators/keyframes.mjs";
import { spring } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/animation/generators/spring/index.mjs";
import { stagger } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/animation/utils/stagger.mjs";
import { clamp } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/clamp.mjs";
import { delayInSeconds } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/delay.mjs";
import { distance, distance2D } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/distance.mjs";
import { interpolate } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/interpolate.mjs";
import { mix } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/mix/index.mjs";
import { pipe } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/pipe.mjs";
import { transform } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/transform.mjs";
import { wrap } from "./external/.pnpm/framer-motion@12.5.0/external/framer-motion/dist/es/utils/wrap.mjs";
import { Motion, motion } from "./components/motion/index.mjs";
import { mountedStates } from "./state/motion-state.mjs";
import { provideMotionConfig, useMotionConfig } from "./components/motion-config/context.mjs";
import { Reorder, ReorderGroup, ReorderItem } from "./components/reorder/index.mjs";
import { LazyMotion } from "./components/lazy-motion/index.mjs";
import { M, m } from "./components/motion/m.mjs";
import { useComputed } from "./value/use-computed.mjs";
import { useCombineMotionValues } from "./value/use-combine-values.mjs";
import { useTransform } from "./value/use-transform.mjs";
import { useTime } from "./value/use-time.mjs";
import { useMotionTemplate } from "./value/use-motion-template.mjs";
import { useMotionValueEvent } from "./value/use-motion-value-event.mjs";
import { useSpring } from "./value/use-spring.mjs";
import { useScroll } from "./value/use-scroll.mjs";
import { useVelocity } from "./value/use-velocity.mjs";
import { useAnimate } from "./animation/hooks/use-animate.mjs";
import { useAnimationControls } from "./animation/hooks/use-animation-controls.mjs";
import { useReducedMotion } from "./animation/hooks/use-reduced-motion.mjs";
import { createContext } from "./utils/createContext.mjs";
import { isMotionValue } from "./utils/motion-value.mjs";
import { useInView } from "./utils/use-in-view.mjs";
import { useAnimationFrame } from "./utils/use-animation-frame.mjs";
import { millisecondsToSeconds, secondsToMilliseconds } from "./utils/time-conversion.mjs";
import { getContextWindow } from "./utils/get-context-window.mjs";
import { useDomRef } from "./utils/use-dom-ref.mjs";
export {
  default4 as AnimatePresence,
  default2 as LayoutGroup,
  LazyMotion,
  M,
  Motion,
  default5 as MotionConfig,
  MotionValue,
  Reorder,
  ReorderGroup,
  ReorderItem,
  default3 as RowValue,
  addScaleCorrector,
  animate,
  animateMini,
  anticipate,
  backIn,
  backInOut,
  backOut,
  cancelFrame,
  circIn,
  circInOut,
  circOut,
  clamp,
  createContext,
  createScopedAnimate,
  cubicBezier,
  delayInSeconds as delay,
  distance,
  distance2D,
  domAnimation,
  domMax,
  easeIn,
  easeInOut,
  easeOut,
  frame,
  frameData,
  getContextWindow,
  hover,
  inView,
  inertia,
  injectLayoutGroup,
  injectMotion,
  interpolate,
  invariant,
  isDragActive,
  isMotionValue,
  keyframes,
  m,
  millisecondsToSeconds,
  mirrorEasing,
  mix,
  motion,
  motionValue,
  mountedStates,
  noop,
  pipe,
  press,
  progress,
  provideLayoutGroup,
  provideMotion,
  provideMotionConfig,
  reverseEasing,
  scroll,
  scrollInfo,
  secondsToMilliseconds,
  spring,
  stagger,
  steps,
  time,
  transform,
  useAnimate,
  useAnimationControls,
  useAnimationFrame,
  useCombineMotionValues,
  useComputed,
  useDomRef,
  useDragControls,
  useInView,
  useLayoutGroup,
  useMotionConfig,
  useMotionTemplate,
  motionValue2 as useMotionValue,
  useMotionValueEvent,
  useReducedMotion,
  useScroll,
  useSpring,
  useTime,
  useTransform,
  useVelocity,
  wrap
};
