// @inspira-ui/plugins/dist/index.mjs
function g(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var s = {}, p = {};
(function(e) {
  Object.defineProperty(e, "__esModule", {
    value: !0
  }), Object.defineProperty(e, "default", {
    enumerable: !0,
    get: function() {
      return n;
    }
  });
  function f(t, r) {
    return {
      handler: t,
      config: r
    };
  }
  f.withOptions = function(t, r = () => ({})) {
    const i = function(a) {
      return {
        __options: a,
        handler: t(a),
        config: r(a)
      };
    };
    return i.__isOptionsFunction = !0, i.__pluginFunction = t, i.__configFunction = r, i;
  };
  const n = f;
})(p);
(function(e) {
  Object.defineProperty(e, "__esModule", {
    value: !0
  }), Object.defineProperty(e, "default", {
    enumerable: !0,
    get: function() {
      return t;
    }
  });
  const f = /* @__PURE__ */ n(p);
  function n(r) {
    return r && r.__esModule ? r : {
      default: r
    };
  }
  const t = f.default;
})(s);
let o = s;
var m = (o.__esModule ? o : { default: o }).default;
const d = /* @__PURE__ */ g(m);
var _ = {
  aqua: /#00ffff(ff)?(?!\w)|#0ff(f)?(?!\w)/gi,
  azure: /#f0ffff(ff)?(?!\w)/gi,
  beige: /#f5f5dc(ff)?(?!\w)/gi,
  bisque: /#ffe4c4(ff)?(?!\w)/gi,
  black: /#000000(ff)?(?!\w)|#000(f)?(?!\w)/gi,
  blue: /#0000ff(ff)?(?!\w)|#00f(f)?(?!\w)/gi,
  brown: /#a52a2a(ff)?(?!\w)/gi,
  coral: /#ff7f50(ff)?(?!\w)/gi,
  cornsilk: /#fff8dc(ff)?(?!\w)/gi,
  crimson: /#dc143c(ff)?(?!\w)/gi,
  cyan: /#00ffff(ff)?(?!\w)|#0ff(f)?(?!\w)/gi,
  darkblue: /#00008b(ff)?(?!\w)/gi,
  darkcyan: /#008b8b(ff)?(?!\w)/gi,
  darkgrey: /#a9a9a9(ff)?(?!\w)/gi,
  darkred: /#8b0000(ff)?(?!\w)/gi,
  deeppink: /#ff1493(ff)?(?!\w)/gi,
  dimgrey: /#696969(ff)?(?!\w)/gi,
  gold: /#ffd700(ff)?(?!\w)/gi,
  green: /#008000(ff)?(?!\w)/gi,
  grey: /#808080(ff)?(?!\w)/gi,
  honeydew: /#f0fff0(ff)?(?!\w)/gi,
  hotpink: /#ff69b4(ff)?(?!\w)/gi,
  indigo: /#4b0082(ff)?(?!\w)/gi,
  ivory: /#fffff0(ff)?(?!\w)/gi,
  khaki: /#f0e68c(ff)?(?!\w)/gi,
  lavender: /#e6e6fa(ff)?(?!\w)/gi,
  lime: /#00ff00(ff)?(?!\w)|#0f0(f)?(?!\w)/gi,
  linen: /#faf0e6(ff)?(?!\w)/gi,
  maroon: /#800000(ff)?(?!\w)/gi,
  moccasin: /#ffe4b5(ff)?(?!\w)/gi,
  navy: /#000080(ff)?(?!\w)/gi,
  oldlace: /#fdf5e6(ff)?(?!\w)/gi,
  olive: /#808000(ff)?(?!\w)/gi,
  orange: /#ffa500(ff)?(?!\w)/gi,
  orchid: /#da70d6(ff)?(?!\w)/gi,
  peru: /#cd853f(ff)?(?!\w)/gi,
  pink: /#ffc0cb(ff)?(?!\w)/gi,
  plum: /#dda0dd(ff)?(?!\w)/gi,
  purple: /#800080(ff)?(?!\w)/gi,
  red: /#ff0000(ff)?(?!\w)|#f00(f)?(?!\w)/gi,
  salmon: /#fa8072(ff)?(?!\w)/gi,
  seagreen: /#2e8b57(ff)?(?!\w)/gi,
  seashell: /#fff5ee(ff)?(?!\w)/gi,
  sienna: /#a0522d(ff)?(?!\w)/gi,
  silver: /#c0c0c0(ff)?(?!\w)/gi,
  skyblue: /#87ceeb(ff)?(?!\w)/gi,
  snow: /#fffafa(ff)?(?!\w)/gi,
  tan: /#d2b48c(ff)?(?!\w)/gi,
  teal: /#008080(ff)?(?!\w)/gi,
  thistle: /#d8bfd8(ff)?(?!\w)/gi,
  tomato: /#ff6347(ff)?(?!\w)/gi,
  violet: /#ee82ee(ff)?(?!\w)/gi,
  wheat: /#f5deb3(ff)?(?!\w)/gi,
  white: /#ffffff(ff)?(?!\w)|#fff(f)?(?!\w)/gi
}, c = _, w = {
  whitespace: /\s+/g,
  urlHexPairs: /%[\dA-F]{2}/g,
  quotes: /"/g
};
function y(e) {
  return e.trim().replace(w.whitespace, " ");
}
function k(e) {
  return encodeURIComponent(e).replace(w.urlHexPairs, O);
}
function P(e) {
  return Object.keys(c).forEach(function(f) {
    c[f].test(e) && (e = e.replace(c[f], f));
  }), e;
}
function O(e) {
  switch (e) {
    case "%20":
      return " ";
    case "%3D":
      return "=";
    case "%3A":
      return ":";
    case "%2F":
      return "/";
    default:
      return e.toLowerCase();
  }
}
function u(e) {
  if (typeof e != "string")
    throw new TypeError("Expected a string, but received " + typeof e);
  e.charCodeAt(0) === 65279 && (e = e.slice(1));
  var f = P(y(e)).replace(w.quotes, "'");
  return "data:image/svg+xml," + k(f);
}
u.toSrcset = function(f) {
  return u(f).replace(/ /g, "%20");
};
var $ = u;
const l = /* @__PURE__ */ g($);
var b = {};
(function(e) {
  Object.defineProperty(e, "__esModule", {
    value: !0
  }), Object.defineProperty(e, "default", {
    enumerable: !0,
    get: function() {
      return n;
    }
  });
  const f = (t) => Object.assign({}, ...Object.entries(t ?? {}).flatMap(([r, i]) => typeof i == "object" ? Object.entries(f(i)).map(([a, v]) => ({
    [r + (a === "DEFAULT" ? "" : `-${a}`)]: v
  })) : [
    {
      [`${r}`]: i
    }
  ])), n = f;
})(b);
const h = /* @__PURE__ */ g(b), C = d(function({ matchUtilities: e, theme: f }) {
  e({
    "bg-grid": (t) => ({
      backgroundImage: `url("${l(
        `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" width="32" height="32" fill="none" stroke="${t}"><path d="M0 .5H31.5V32"/></svg>`
      )}")`
    }),
    "bg-grid-small": (t) => ({
      backgroundImage: `url("${l(
        `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" width="8" height="8" fill="none" stroke="${t}"><path d="M0 .5H31.5V32"/></svg>`
      )}")`
    }),
    "bg-dot": (t) => ({
      backgroundImage: `url("${l(
        `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" width="16" height="16" fill="none"><circle fill="${t}" id="pattern-circle" cx="10" cy="10" r="1.6257413380501518"></circle></svg>`
      )}")`
    })
  }, {
    values: h(f("backgroundColor")),
    type: "color"
  });
}), j = d(({ addBase: e, theme: f }) => {
  const n = h(f("colors")), t = Object.fromEntries(
    Object.entries(n).map(([r, i]) => [`--${r}`, i])
  );
  e({
    ":root": t
  });
}), E = d(function(e) {
  j.handler(e), C.handler(e);
});
export {
  C as addBackgrounds,
  j as addVariablesForColors,
  E as setupInspiraUI
};
